streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -368,50 +378,134 @@ void handleLostAll() {
                     }
                 } catch (final TaskIdFormatException e) {
-                    // there may be some unknown files that sits in the same directory,
-                    // we should ignore these files instead trying to delete them as well
+                    // ignore any unknown files that sit in the same directory
+                }
+            }
+        }
+    }
+
+    /**
+     * Attempts to release the lock for the passed in task's directory. It does not remove the task from
+     * {@code lockedTaskDirectories} so it's safe to call during iteration, and should be idempotent.
+     */
+    private RuntimeException releaseTaskDirLock(final TaskId taskId) {
+        try {
+            stateDirectory.unlock(taskId);
+        } catch (final IOException e) {
+            log.error("Failed to release the lock for task directory {}.", taskId);
+            return new StreamsException(String.format("Unable to unlock task directory %s", taskId), e);
+        }
+        return null;
+    }
+
+    /**
+     * We must release the lock for any unassigned tasks that we temporarily locked in preparation for a
+     * rebalance in {@link #tryLockForAllTaskDirectories()}.
+     */
+    private void releaseLockedUnassignedTaskDirectories() {
+        final AtomicReference<RuntimeException> firstException = new AtomicReference<>(null);
+
+        final Iterator<TaskId> taskIdIterator = lockedTaskDirectories.iterator();
+        while (taskIdIterator.hasNext()) {
+            final TaskId id = taskIdIterator.next();
+
+            if (!tasks.containsKey(id)) {
+                final RuntimeException unlockException = releaseTaskDirLock(id);
+                if (unlockException == null) {
+                    taskIdIterator.remove();
+                } else {
+                    firstException.compareAndSet(null, unlockException);
                 }
             }
         }
 
-        return locallyStoredTasks;
+        final RuntimeException fatalException = firstException.get();
+        if (fatalException != null) {
+            throw fatalException;
+        }
+    }
+
+    private long sumOfChangelogOffsets(final Map<TopicPartition, Long> changelogOffsets) {
+        long offsetSum = 0L;
+        for (final Map.Entry<TopicPartition, Long> changelogEntry : changelogOffsets.entrySet()) {
+            final TopicPartition changelog = changelogEntry.getKey();
+            final long offset = changelogEntry.getValue();
+
+            if (offset < 0L) {
+                if (offset == -1L) {
+                    log.debug("Skipping unknown offset for changelog {}", changelog);
+                } else {
+                    log.warn("Unexpected negative offset {} for changelog {}", offset, changelog);
+                }
+                continue;
+            }
+
+            offsetSum += offset;
+            if (offsetSum < 0) {
+                return Long.MAX_VALUE;
+            }
```

Comment:
That's fair, and I'm _pretty_  sure Java does guarantee that. However I still have two concerns:
1. If it can overflow once, it can overflow twice. If it wraps back around to positive offsets we won't be able to distinguish an instance that's caught up to extremely large offsets from one that's just started
2. We'd have to protect `Task.LATEST_OFFSET`, which would be awkward (though admittedly not terribly so). But, what if we need to introduce new sentinels?

Reply:
This is hard to reason about in a vacuum, so here's some rough numbers to give things context. I found this quote that gives perspective on what you can do with one partition whose offsets are stored as long:

> If you write 1TB a day, you can keep going for about 4 million days

This corresponds to roughly 11.5MB/s for 11,000 years. Of course a Streams app will have more than one store per partition. Most topologies probably cap out at around 5-10 stores/partition, but let's take an extreme example and say you have 1000 stores/partition.

With this setup you could process **11.5MB/s for 11 years**, assuming you hit no other limits -- and I'm pretty skeptical an app with 1000+ stores per instance could smoothly reach 11.5MB/s to begin with.

So with that in mind, do we think a long is sufficient? cc/ @vvcephei

Reply:
I was looking over the history of this discussion to see what Bruno's proposal actually was, and I couldn't find it. This? https://github.com/apache/kafka/pull/8246/#discussion_r390563165
But I didn't see anything about not using a long. I think that comment was that you can just let the overflow happen, and then when you compute lag, the end-offsets would also have overflown, and they'll bear a predictable relationship to each other, which is true.

Then again, Sophie makes a good point that overflowing Long with offsets, even with multiple stores per task, is pretty hard in practice. In that case, it would still be nice to detect it and just throw an exception, rather than allowing nonsense to propagate through the application logic. I think 1000 stores in a single task is probably on the high side, but who knows... I think 11 years is kind of a long time, but then again, I've personally worked with databases that had been running continuously for longer than that.

Reply:
@vvcephei Not 100% what you're getting at in the first paragraph, but I don't think Bruno was suggesting to not use a long, just to let it overflow. My concern was that if we really think it's possible to overflow past `Long.MAX_VALUE` then we should also be concerned about it doing so again and wrapping back to positive offsets again. This would cause more serious degradation to the assignment algorithm since really-caught-up clients would be indistinguishable from just-started ones.
On the other hand, pinning overflow to `Long.MAX_VALUE` will at worst cause the algorithm to reduce to the current one as all clients will appear caught-up.

Reply:
Not to mention, unless they do have a truly unreasonable number of stores per partition AND the app manages to keep up with the input topics, it's pretty likely they would be doomed to hit the MAX_VALUE on at least one partition's offset not long after hitting this on the offset sum (if not before). And there's nothing we can do about that.

Reply:
@ableegoldman Good point about not be able to distinguish a new instance from a caught-up instance when the offset sums get positive again (I do not want to imagine how long restoration would take in that case ;-)). Let's pin the offset at `MAX_VALUE` in the rare case of an overflow and issue a WARN log message. OK?

Reply:
One thing to get out of the way, you can't overflow back to positive again with addition: `MAX_VALUE + MAX_VALUE = -2`, so as long as you check after each addition operation, you're 100% covered. If you really want to be paranoid that maybe some future version of Java would change the relationship among max, min, and zero such that `max*2 is > 0` (which I think is unlikely), you could instead check that the result of adding two numbers is greater than both operands.

I'm ok with a saturation strategy, and just pinning at MAX_VALUE. We should do the same when computing end offsets as when computing current position. Although, I guess that means that by the time a bunch of tasks saturate their positions (gets pinned to MAX_VALUE), the assignment would just throw up its hands and say they're all caught up, since they all have the same value as the end-offset (which also would have saturated). This seems like a bummer, but it's either that or crash. Since the result isn't incorrect processing, just a potentially suboptimal assignment, I think saturation is a fine approach in conjunction with a warning.

Reply:
Note that we can always bump the protocol version in 10 years, and start encoding the offset sums as a byte[], BigInt, etc

Reply:
I'll add the warning for now -- presumably if someone actually hits this on a task they'll ask about the warning and/or submit a ticket.

And that's how we'll know it's time to bump the protocol 😉